src/dhtnode/storage/StorageEngineStepIterator.d

```diff
@@ -148,14 +176,25 @@ public class StorageEngineStepIterator
         verify(this.storage !is null,
             typeof(this).stringof ~ ".next: storage not set");
 
-        if (this.started)
-            this.storage.getNextKey(this.current_key, this.key_buffer,
-                this.current_key);
-        else
+        bool more;
+        with ( State ) switch ( this.state )
         {
-            this.started = true;
-            this.storage.getFirstKey(this.key_buffer, this.current_key);
+            case Init:
+                this.state = Started;
+                more = this.storage.getFirstKey(this.current_key);
+                break;
+            case Started:
+                more = this.storage.getNextKey(this.current_key,
+                    this.current_key);
+                break;
+            case Finished:
+                break;
+            default:
+                verify(false);
         }
+
+        if ( !more )
+            this.state = State.Finished;
```

Comment:
This thing with the `State` and refactoring here could be a separate commit, right?

Reply:
Hmm... yeah, that commit is a bit weird. Just trying to figure out what's happening there :/

Reply:
Added a bit of explanation in the commit message.